src/image.rs

```diff
@@ -0,0 +1,221 @@
+    ///
+    /// # Example
+    /// ```~/.injesh/images/alpine/3.15/rootfs```
+    fn rootfs_path(&self) -> PathBuf {
+        PathBuf::from(&format!("{}/{}", self.image_base_path().display(), ROOTFS))
+    }
+
+    /// ローカルにrootfsイメージがあるかどうか調べる
+    ///
+    /// # Example
+    /// ```ignore
+    /// image.search_image();
+    /// ```
+    pub fn search_image(&self) -> Result<(), Box<dyn std::error::Error>> {
+        if self.rootfs_hash_path().exists() {
+            return Ok(());
+        }
+        Err(Error::ImageNotFound)?
+    }
+
+    /// ローカルにあるイメージが最新のものかどうかを調べる
+    ///
+    /// # Example
+    /// ```ignore
+    /// image.image_is_newes();
+    /// ```
+    pub fn check_rootfs_newest(&self) -> Result<bool, Box<dyn std::error::Error>> {
+        self.downloader
+            .check_rootfs_newest(&self.rootfs_hash_path())
+    }
+
+    /// rootfsイメージをダウンロードする
+    /// 既にrootfsが存在する場合、削除してからダウンロードする
+    /// そのため存在するかどうか、最新かどうかを確認してから
+    /// 呼び出す必要がある
+    ///
+    /// # Example
+    /// ```ignore
+    /// image.download_image();
+    /// ```
+    pub fn download_image(&self) -> Result<(), Box<dyn std::error::Error>> {
+        // ローカルにあるrootfsが最新のものかどうかを確認
+        if Path::new(&self.rootfs_hash_path()).exists() {
+            match self.check_rootfs_newest() {
+                Ok(newest_flg) => match newest_flg {
+                    true => return Ok(()),
+                    false => { /* do nothing */ }
+                },
+                Err(_) => { /* do nothing */ }
+            }
+        }
+
+        self.setup_rootfs_directory()?;
+
+        self.downloader
+            .download_rootfs(&self.downloaded_rootfs_path())?;
+
+        self.downloader
+            .download_rootfs_hash(&self.rootfs_hash_path())?;
+
+        // ダウンロードしたrootfsを解凍
+        let tar_xz = File::open(self.downloaded_rootfs_path())?;
+        let tar = XzDecoder::new(tar_xz);
+        let mut archive = Archive::new(tar);
+        archive.unpack(self.rootfs_path())?;
+
+        // ダウンロードしたtarファイルを削除
+        fs::remove_file(self.downloaded_rootfs_path())?;
+
+        Ok(())
+    }
+
+    /// rootfsイメージを格納するディレクトリを生成する
+    /// 既に```~/.injesh/distri/version/rootfs```以下が存在する場合削除する
+    ///
+    /// # Example
+    /// ```ignore
+    /// image.setup_rootfs_directory();
+    /// ```
+    fn setup_rootfs_directory(&self) -> Result<(), Box<dyn std::error::Error>> {
+        // rootfsのディストリビューションを表すディレクトリのチェック
+        if !Path::new(&format!("{}/{}", self.user().images(), self.distribution())).exists() {
+            fs::create_dir(format!("{}/{}", self.user().images(), self.distribution()))?
+        }
+
+        // rootfsのバージョンを表すディレクトリのチェック
+        if !Path::new(self.image_base_path()).exists() {
+            fs::create_dir(self.image_base_path())?
+        }
+
+        if !Path::new(&self.rootfs_path()).exists() {
+            fs::create_dir(&self.rootfs_path())?;
+        } else {
+            fs::remove_dir_all(&self.rootfs_path())?;
+        }
```

Comment:
ディレクトリがない，or バージョンが古い(のはこの関数を呼び出す前に確認しているが，誤って確認を逃し，最新版であっても消してしまうことを防ぐ意味でも，この関数に入れたほうが便利では？) なら強制的に(再)DLするということで
また，実際にDLするときに rootfs ディレクトリが作成されるため，ここでは作らないようにする

```suggestion
        fs::create_dir_all(self.image_base_path())?

        // rootfs があるなら古いバージョンと見做して消去する
        if Path::new(&self.rootfs_path()).exists() {
            fs::remove_dir_all(&self.rootfs_path())?;
        }
```

Reply:
> 実際にDLするときに rootfs ディレクトリが作成されるため

というのはアーカイブをダウンロードして展開した時ということ？

Reply:
そう
https://github.com/higuruchi/injesh/pull/26/files#diff-9e8d03b9988bc5a6f26a988a9d956e3dbaf95ea6d9c9eb55676c311f76b32140R187
ここで作られると思っている

もしかして作られない？
なら，elseを加えて，rootfsがないときはディレクトリを作成する
とする？

Reply:
作られないです

Reply:
りょ

Reply:
create_dir_allは便利そうなので利用します。